File: dotnet/ThirtyFour.UI/Table/ColumnHeaderCellFinder.cs
using OpenQA.Selenium;
using System;
using System.Collections.Generic;
using System.Linq;

namespace ThirtyFour.UI.Table
{
    /// <summary>
    /// Finds a cell by correlating the table headers with a value to identify the correct table row
    /// </summary>
    public class ColumnHeaderCellFinder : ICellFinder
    {

        protected string findColumnHeader;
        protected string findValue;
        protected string returnColumnHeader;
        protected IWebElement tableElement;

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="findColumnHeader">The header of the column used to find the correct table row</param>
        /// <param name="findValue">The value which identifes the correct table row (within the find column)</param>
        /// <param name="returnColumnHeader">The column header used to identify the correct cell</param>
        public ColumnHeaderCellFinder(string findColumnHeader, string findValue, string returnColumnHeader)
        {
            this.findColumnHeader = findColumnHeader;
            this.findValue = findValue;
            this.returnColumnHeader = returnColumnHeader;
        }


        /// <summary>
        /// Returns the zero based index/position of a column with the matching header
        /// </summary>
        /// <param name="columnHeader">The column header</param>
        /// <returns>The index of the column</returns>
        protected int GetColumnIndex(string columnHeader)
        {
            IReadOnlyCollection<IWebElement> headerCells = this.tableElement.FindElements(By.TagName("th"));
            for (int i = 0; i < headerCells.Count; i++)
            {
                if (headerCells.ElementAt(i).Text.Equals(columnHeader))
                {
                    return i;
                }
            }

            throw new NoSuchElementException(String.Format("Unable to find column with header text '{0}'", columnHeader));
        }

        /// <summary>
        /// Finds and returns a specific table cell
        /// </summary>
        /// <param name="tableElement">the element representing the html table</param>
        /// <returns>an element representing the tabel cell</returns>
        public IWebElement FindCell(IWebElement tableElement)
        {
            this.tableElement = tableElement;
            int findColumnIndex = this.GetColumnIndex(this.findColumnHeader);
            int returnColumnIndex = this.GetColumnIndex(this.returnColumnHeader);

            IReadOnlyCollection<IWebElement> rows = tableElement.FindElements(By.CssSelector("tbody tr"));
            foreach (IWebElement row in rows)
            {
                IReadOnlyCollection<IWebElement> cells = row.FindElements(By.TagName("td"));
                if (cells.ElementAt(findColumnIndex).Text.Equals(this.findValue))
                {
                    return cells.ElementAt(returnColumnIndex);
                }
            }

            throw new NoSuchElementException(String.Format("Unable to find row containing find value '{0}'", findValue));
        }
    }
}

File: java/src/main/java/com/github/danhumphrey/thirtyfour/ui/table/ColumnHeaderCellFinder.java
package com.github.danhumphrey.thirtyfour.ui.table;

import java.util.List;

import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;

/**
 * 
 * Finds a row and cell by correlating the table headers and a cell value 
 */
public class ColumnHeaderCellFinder implements CellFinder {

	protected String findColumnHeader;
	protected String findColumnValue;
	protected String returnColumnHeader;
	
	protected WebElement tableElement;
	
	/**
	 * The default constructor
	 * 
	 * @param findColumnHeader The header of the column in which to look for the findColumnValue
	 * @param findColumnValue The value to be located within the find column - this identifies the correct row
	 * @param returnColumnHeader The header of the column to return
	 */
	public ColumnHeaderCellFinder(String findColumnHeader, String findColumnValue, String returnColumnHeader) {
		this.findColumnHeader = findColumnHeader;
		this.findColumnValue = findColumnValue;
		this.returnColumnHeader = returnColumnHeader;
	}
	
	/**
	 * Returns the zero based index/position of a column
	 * 
	 * @param columnHeader the header of the column to locate
	 * @return the position of the column if found
	 * @throws NoSuchElementException if the column can't be found
	 */
	protected int getColumnIndex(String columnHeader) throws NoSuchElementException {
		List<WebElement> headerCells = this.tableElement.findElements(By.tagName("th"));
		for(int i = 0; i < headerCells.size(); i++) {
			if(headerCells.get(i).getText().equals(columnHeader)) {
				return i;
			}
		}
		throw new NoSuchElementException(String.format("Unable to find column with header text %s", columnHeader));

	}
	
	@Override
	public WebElement findCell(WebElement tableElement) throws NoSuchElementException {
		
		this.tableElement = tableElement;
		int findColumnIndex = this.getColumnIndex(this.findColumnHeader);
		int returnColumnIndex = this.getColumnIndex(this.returnColumnHeader);
		
		List<WebElement> rows = tableElement.findElements(By.cssSelector("tbody tr"));
		for(WebElement row : rows) {
			List<WebElement> cells = row.findElements(By.tagName("td"));
			if(cells.get(findColumnIndex).getText().equals(this.findColumnValue)) {
				return cells.get(returnColumnIndex);
			}
		}
		
		throw new NoSuchElementException(String.format("Unable to find cell with text %s", findColumnValue));
	}

}
